Sources/Runtime/arc.cpp

```diff
@@ -157,28 +159,58 @@ struct RefCounted {
 
 private:
   /**
+   Returns whether the value inside the \c RefCountBox is in the
+   process of being deallocated
+   @note This function *must* be called with a locked \c mutex.
+   */
+  bool isDeallocating() {
+    return box->retainCount & 0x1;
+  }
+  
+  /**
+   Returns the actual retain count of the \c RefCountBox
+   @note This function *must* be called with a locked \c mutex.
+   */
+  bool lockedRetainCount() {
+    return ((box->retainCount & ~ 0x1)) >> 1;
+  }
+  
+  /**
    Deallocates the value inside a \c RefCountBox.
    @note This function *must* be called with a locked \c mutex.
    */
   void dealloc() {
     CHECK_VALID_BOX();
+#ifdef DEBUG_ARC
+    if (box->deallocated) {
+      DEBUG_ARC_LOG("dealloc on an already deallocated object");
+      box->mutex.unlock();
+      return;
+    }
+#endif
     if (box->retainCount > 0) {
       trill_fatalError("object deallocated with retain count > 0");
     }
 
     DEBUG_ARC_LOG("deallocating");
 
     if (box->deinit != nullptr) {
+      box->retainCount = 1;
       box->deinit(value);
+      if (box->retainCount != 1) {
+        trill_fatalError("attempted to resurrect zombie during deinit... this is an apocalypse-safe language");
+      }
     }
 
+    box->retainCount = 0; // clear the deallocating flag
+    
     // Explicitly unlock the mutex before deleting the box
     box->mutex.unlock();
 
 #ifdef DEBUG_ARC
     box->deallocated = true;
 #else
-    free(box);
+    delete box;
```

Comment:
I don't think that's true. C++ has no way of knowing to free the entire buffer that box was allocated into (unless we're relying on an implementation detail of `delete` using free and asking the heap table for the buffer size)

From the googling I've done, it looks like the correct thing here, as we allocated this buffer larger than the object size, is

```c++
box->~RefCountBox();
free(box);
```